File: ruby/lib/resp_bench/command/command.rb
# frozen_string_literal: true

module RespBench
  module Command
    # Result of a command execution, used for metrics collection.
    CommandResult = Struct.new(:command_name, :latency_micros, :success, keyword_init: true) do
      def success?
        success
      end
    end

    # Abstract base class for benchmark commands.
    class Command
      attr_reader :weight, :name

      def initialize(config)
        @weight = config.weight
        @name = config.command.upcase
        @data_size_bytes = config.data_size_bytes
      end

      # Execute the command and return a result for metrics
      # @param client [RespBench::Client::BenchmarkClient] the client
      # @param key_generator [RespBench::Engine::KeyGenerator] key generator
      # @return [CommandResult] the result with latency and success status
      def execute(client, key_generator)
        raise NotImplementedError, "#{self.class} must implement #execute"
      end
    end
  end
end

File: python/src/resp_bench/command/command.py
"""Command base class and the metrics-facing result struct."""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Optional

from ..client.benchmark_client import AsyncBenchmarkClient
from ..config.command_config import CommandConfig


@dataclass
class CommandResult:
    command_name: str
    latency_micros: int
    success: bool


class Command(ABC):
    def __init__(self, config: CommandConfig) -> None:
        self.weight = config.weight
        self.name = config.command.upper()
        self._data_size_bytes = config.data_size_bytes

    @abstractmethod
    async def execute(self, client: AsyncBenchmarkClient, key: Optional[str]) -> CommandResult:
        """Execute against ``client`` (using ``key`` where applicable)."""
